### PR TITLE
use ParamSpec for @contextmanager

### PR DESCRIPTION
**Pull request by JelleZijlstra**
#4827 doesn't have all boxes checked yet, but I want to see what CI has to say about this.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
CI says that mypy needs more work to support this.

File: stdlib/contextlib.pyi [hunk 1]
```diff
@@ -23,17 +23,19 @@ _T = TypeVar("_T")
...
+_P = ParamSpec("_P")
 
 _ExitFunc = Callable[[Optional[Type[BaseException]], Optional[BaseException], Optional[TracebackType]], bool]
 _CM_EF = TypeVar("_CM_EF", ContextManager[Any], _ExitFunc)
 
 class _GeneratorContextManager(ContextManager[_T_co]):
     def __call__(self, func: _F) -> _F: ...
 
-def contextmanager(func: Callable[..., Iterator[_T]]) -> Callable[..., _GeneratorContextManager[_T]]: ...
+# type: ignore to deal with incomplete ParamSpec support in mypy
```
Comment (srittau) [inline stdlib/contextlib.pyi:34]: This might throw mypy, even if it complains about line 35. Just remove the ":" maybe.

Comment (JelleZijlstra) [reply stdlib/contextlib.pyi]: Oops, you're right

**Comment by srittau** [comment]
This is looking promising.

**Comment by JelleZijlstra** [comment]
Interesting, I thought the mypy errors were blocking errors (preventing it from analyzing the rest of the file), but I suppose not.

**Review by srittau — APPROVED** [review]

**Pull request merged**